### ISSUE TITLE
Adaptive usage of memory during training inference

### ISSUE THREAD

**Issue by eaedk** [pytorch/pytorch#45295]
## 🚀 Feature

Better management of GPU and RAM memory to train CNN models with larger image size.

## Motivation

Find a way to train the model (CNN) with better frame size (720 * 720) on GPUs like the Nvidia K80 by splitting the inference process.

## Pitch

- Save a part, or some parameters of the models on the RAM in order to allow more data to be passed to CNNs.
- Or infer on a batch by cutting it into several small sets so as not to have memory allocation errors.

**Comment by eaedk** [comment]
You can ask questions if you don't fully understand my issue

**Comment by mruberry** [comment]
Thank you for filing this issue. This request has come up before (we may have an existing issue for it, actually). I would be interested to see a prototype of this behavior as well as its impact on performance. How complicated would a system like this have to be before it would outperform only running part of the network on CUDA?

**Comment by eaedk** [comment]
I don’t really have answers. I never went deeper in the source code of PyTorch. What I want is a way to avoid memory error in three cases : 
1 - when models are too large for gpu’s memory;
2 - when we want to pass big images;
3 - when gpu’s memory was use before by another models.

**Comment by eaedk** [comment]
Hypothesis for the three points : 
1 - Use RAM to save some model’s parameters or intermediate computing results ;
2 - An auto gradient accumulator which compute the number of images that could feed the model;
3 - A gpu’s memory cleaner.

**Comment by albanD** [comment]
Hi,

We already have solutions for a lot of these:
- You can use `utils.checkpoint` to recompute the temporary buffers instead of saving them.
- You can split up your batch to reduce overall memory without changing the behavior (see [this](https://discuss.pytorch.org/t/why-do-we-need-to-set-the-gradients-manually-to-zero-in-pytorch/4903/20) forum post for details).
- The GPU memory is available for use for new Tensors as soon as nothing references it. So no need for a special cleaner. If the GPU is shared by multiple programs, you can use `cuda.empty_cache()` to make that memory available to other process as well.
- You can use DistributedDataParallel to split your batch on multiple GPUs if you have them.

We don't have any utility to move things to RAM as for regular neural nets, recomputing the values on the GPU is actually faster than doing two copies from between the the CPU and GPU.
